flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/internals/publisher/fanout/FanOutShardSubscriber.java

```diff
@@ -272,8 +289,6 @@ private boolean consumeAllRecordsFromKinesisShard(
 
 		private final CountDownLatch waitForSubscriptionLatch;
 
-		private final Object lockObject = new Object();
```

Comment:
Making sure that I understand the new error throwing mechanism, from the looks of the code:

- Exceptions are now passed via an atomic reference, instead of being put into the queue
- Because of that, we no longer need to synchronize access to the queue (because it'll only ever be accessed serially in the `onNext` and `onComplete` methods). Therefore, this lock object is no longer needed.
- Also because of that, we can now simply block indefinitely when appending a subscription event to the queue, because we no longer need to let error events have a chance to take over.

What I am failing to understand is why we can now only handle errors _after_ the queue is fully consumed.

Reply:
Yes you are correct. 

The order of processing that dictates we handle errors after the queue is fully consumed is similar to before. When an exception is thrown in the network thread, an error is always the last thing the consumer (record publisher) receives. Now we are not discarding batches to deliver that exception. 

A better way to look at it is, what does handling the error actually mean? For this we would reacquire the EFO subscription to continue consuming data. We do not want to do this until we have processed all the records that have already arrived. 

Let me know if this is still not clear.